File: source/com/barrybecker4/simulation/habitat1/creatures/CreatureType.java
/** Copyright by Barry G. Becker, 2000-2011. Licensed under MIT License: http://www.opensource.org/licenses/MIT  */
package com.barrybecker4.simulation.habitat1.creatures;

import java.awt.*;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Everything we need to know about a creature.
 * There are many different sorts of creatures.
 *
 * Add more creatures like sheep, fox, chickens, ants, vulture,
 *
 * @author Barry Becker
 */
public enum CreatureType {
                                        // size,   mSpeed,  normpeed, gest, starve,  nutr
    GRASS("grass", new Color(40, 255, 20),    2.0,    0.0,      0.0,    11,    42,     1),
    WILDEBEEST(  "cow",   new Color(70, 60, 100),   15.0,    0.002,    0.0001,  25,    94,    10),
    RAT(  "rat",   new Color(140, 105, 20),   2.0,    0.005,    0.002,   8,    30,     1),
    CAT(  "cat",   new Color(0, 195, 220),    5.0,    0.01,     0.004,  15,    58,     4),
    LION("lion",   new Color(240, 200, 20),   9.0,    0.02,    0.008,   21,    83,     6);


    private String name;
    private Color color;
    private double size;
    private double normalSpeed;
    private double maxSpeed;

    /** when numDaysPregnant = this then reproduce */
    private int gestationPeriod;

    /** when hunger >= this then die. */
    private int starvationThreshold;

    private int nutritionalValue;

    private static final Map<CreatureType, List<CreatureType>> predatorMap = new HashMap<CreatureType, List<CreatureType>>();
    private static final Map<CreatureType, List<CreatureType>> preyMap = new HashMap<CreatureType, List<CreatureType>>();


    static {
        // eaten by relationship
        predatorMap.put(GRASS, Arrays.asList(WILDEBEEST, RAT));
        predatorMap.put(WILDEBEEST, Arrays.asList(LION));
        predatorMap.put(RAT, Arrays.asList(CAT, LION));
        predatorMap.put(CAT, Arrays.asList(LION));
        predatorMap.put(LION, Collections.<CreatureType>emptyList());

        for (CreatureType creature : values()) {
            List<CreatureType> preys = new ArrayList<CreatureType>();

            for (CreatureType potentialprey : values()) {
                List<CreatureType> preds = predatorMap.get(potentialprey);
                if (preds.contains(creature)) {
                    preys.add(potentialprey);
                }
            }
            preyMap.put(creature, preys);
        }
    }

    CreatureType(String name, Color color, double size, double normlSpeed,
                 double maxSpeed, int gestationPeriod, int starvationTheshold, int nutritionalValue)  {
        this.name = name;
        this.color = color;
        this.size = size;
        this.normalSpeed = normlSpeed;
        this.maxSpeed = maxSpeed;
        this.gestationPeriod = gestationPeriod;
        this.starvationThreshold = starvationTheshold;
        this.nutritionalValue = nutritionalValue;
    }

    public String getName() {
        return name;
    }

    public Color getColor() {
        return color;
    }

    public double getSize() {
        return size;
    }

    public void setSize(double s) {
        size = s;
    }

    public double getNormalSpeed() {
        return normalSpeed;
    }

    public void setNormalSpeed(double s) {
        normalSpeed = s;
    }

    public double getMaxSpeed() {
        return maxSpeed;
    }

    public void setMaxSpeed(double s) {
        maxSpeed = s;
    }

    public int getGestationPeriod() {
        return gestationPeriod;
    }

    public void setGestationPeriod(int g) {
        gestationPeriod = g;
    }

    public int getStarvationThreshold() {
        return starvationThreshold;
    }

    public void setStarvationThreshold(int value) {
        starvationThreshold = value;
    }

    public int getNutritionalValue() {
        return nutritionalValue;
    }

    public void setNutritionalValue(int value) {
        nutritionalValue = value;
    }

    public List<CreatureType> getPredators() {
        return predatorMap.get(this);
    }

    public List<CreatureType> getPreys() {
        return preyMap.get(this);
    }

    public String toString() {
        return getName();
    }

    /** for testing */
    public static void main(String[] args) {
        System.out.println("preyMap = " + preyMap);
    }
}

File: scala-source/com/barrybecker4/simulation/habitat/creatures/CreatureType.scala
// Copyright by Barry G. Becker, 2016-2023. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.simulation.habitat.creatures

import java.awt.Color
import com.barrybecker4.simulation.habitat.creatures.CreatureType._


/**
  * Everything we need to know about all the different sorts of creatures.
  * Add more creatures like sheep, fox, chickens, ants, vulture, etc.
  * This is probably the best pattern to follow for enums with properties and methods.
  * @author Barry Becker
  */
object CreatureType  {

  val GRASS = new CreatureType("grass", new Color(120, 180, 10), 6.0,
    0.0,    0.0,    15, 5,  280,
    47,     70, 0)
  val COW = new CreatureType("cow", new Color(140, 140, 240),    20.0,
    0.007,  0.002,  39,  40,  390,
    67,     110, 2,  0.5)
  val RAT = new CreatureType("rat", new Color(150, 125, 100),    3.0,
    0.007,  0.002,   14, 11, 120,
    27,     12, 1, 0.7)
  val CAT = new CreatureType("cat", new Color(190, 135, 70),     6.0,
    0.024,  0.005,  21,  30,  310,
    33,     25, 2,0.2)
  val LION = new CreatureType("lion", new Color(180, 160, 30),   11.0,
    0.03, 0.01,  29, 40,  330,
    36,     50, 4,0.3)

  val VALUES = Array(GRASS, COW, RAT, CAT, LION)

  private val preyToPredatorsMap: Map[CreatureType, Set[CreatureType]] = Map(
    GRASS -> Set(COW, RAT),
    COW -> Set(LION),
    RAT -> Set(CAT, LION),
    CAT -> Set(LION),
    LION -> Set[CreatureType]() // king of the jungle, top of the food chain
  )

  private var predatorToPreysMap: Map[CreatureType, Set[CreatureType]] = Map[CreatureType, Set[CreatureType]]()
  for (creature <- VALUES) {
    var preys = Set[CreatureType]()
    for (potentialPrey <- VALUES) {
      val preds: Set[CreatureType] = preyToPredatorsMap(potentialPrey)
      if (preds.contains(creature)) preys += potentialPrey
    }
    predatorToPreysMap += (creature -> preys)
  }
}

class CreatureType(val name: String, val color: Color,
                   var size: Double, var maxSpeed: Double, var normalSpeed: Double,
                   var gestationPeriod: Int, var timeToMaturity: Int, var maxAge: Int,
                   var starvationThreshold: Int, var nutritionalValue: Int, var eatRate: Int,
                   var flockTendancy: Double = 0, var spawnRate: Int = 0) {

  def getPreys: Set[CreatureType] = CreatureType.predatorToPreysMap(this)
  override def toString: String = name
}
